File: ContainerWithMostWater.java

public class ContainerWithMostWater {

	public int maxArea(int[] height) {
		int left = 0;
		int right = height.length - 1;
		int area = left * right;

		while (left < right) {
			area = Math.max(area, (right - left) * Math.min(height[left], height[right]));
			if (height[left] < height[right]) {
				left++;
			} else {
				right--;
			}
		}
		return area;
	}
}

File: ContainerWithMostWater.js
var maxArea = function(heights) {
    let max = 0;
    let left = 0;
    let right = heights.length - 1;
    
    // while pointers are positioned correctly
    while (left < right) {
        // calculate area for current container
        const minHeight = Math.min(heights[left], heights[right]);
        const area = minHeight * (right - left);
        
        if (area > max) {
            max = area;
        }
        
        // if left height is the limiting factor, the only way area can be increased is if left height is increased, so increment left pointer
        if (heights[left] < heights[right]) {
            left++;
        } 
        // else, increment right pointer
        else {
            right--;
        }
    }
    
    return max;
}